File: 242-valid-anagram/valid-anagram.java
class Solution {
    public boolean isAnagram(String s, String t) {

        if(s.length() != t.length()) return false;

        int[] count = new int[26];

        for(int i = 0; i < s.length(); i++) {
            count[s.charAt(i) - 'a']++;
            count[t.charAt(i) - 'a']--;
        }

        for(int c : count) {
            if(c != 0) return false;
        }

        return true;
    }
}

// Notes
// Identify Problem | What is anagram?
// Anagram = any word that can be rearranged to make another word, not just reverse.

// Solution
// Utilize an arr[26] to track letters and their freqency;

// As we loop string everytime we see a letter we increment their freq
// Concurrently we do the same for the second string but we dec
// if true anagram, all values should negate out and be 0;

// thus utilize a foreach and check at the end

File: 242-valid-anagram/valid-anagram.py
from collections import defaultdict

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:

        if len(s) != len(t):
            return False

        freq = [0] * 26

        for c in s:
            freq[ord(c) - ord('a')] += 1

        for c in t:
            freq[ord(c) - ord('a')] -= 1
            
        for char_freq in freq:
            if char_freq != 0:
                return False

        return True

    #Solution 2 use arrayFreq based
    # slightly faster than map

    #Utilze ord() to get unicode for letter

    # Solution 1 | dicts
        # anagram_dict = defaultdict(int)

        # #first loop | count char in first word
        # for c in s:
        #     anagram_dict[c] += 1
        
        # #second loop | count/subtract char from second word
        # for c in t:
        #     anagram_dict[c] -= 1

        # #char count should cancel out and == 0
        # #if not == 0, not anagram
        # for count in anagram_dict.values():
        #     if count != 0:
        #         return False

        # #default true
        # return True